### ISSUE TITLE
New Bill Submission

### ISSUE THREAD

**Issue by adamnew123456** [markjm/426project#5]
Discussion thread.

**Comment by adamnew123456** [comment]
I've looked into this a bit, and I have successfully gotten a virtualenv up and running on classroom.cs.unc.edu and managed to get a configuration close to what I'll need for the scraping side. I haven't done a complete test yet, I'll update with results when I do.

**Comment by adamnew123456** [comment]
I've got some basic code to work between the existing scraping code I have, and the database.

One odd effect is that it is returning a bunch of superfluous bills, since the RSS feed is full of junk. I'll have to look and see if there's anything I can do to filter them out.

**Comment by adamnew123456** [comment]
The above issue seems to be largely fixed - the script looks like it's doing a good job of getting rid of junk entries. I'll see what needs to be done about scheduling it and exposing pending bills via REST.

**EDIT** One question - to avoid having a latency spike when a person forces a request update, I could have the server send out an extra 'update_url' in the JSON, and the frontend could call that URL via AJAX to arrange for the server to run the update without affecting the page-load. Would this be something you'd be willing to do?

**Comment by markjm** [comment]
If you think that would be better, then definitely.

Let me know if you need any help on any thing else.

**Comment by markjm** [comment]
Also, if more times are available, do you want to move back our presentation? I don't particularly mind either way

**Comment by adamnew123456** [comment]
It depends on what my work schedule for exams looks like (which I probably won't have until at some point on Monday) - depending on what that looks like, we might be better off moving it to one of the slots Friday morning. 

A lot of the larger free spots (early Saturday and Tuesday afternoon) conflict with an exam - it'll definitely be Friday, I'm just not sure when.

**Comment by adamnew123456** [comment]
OK, so with AJAX updating the response is going to look a bit different:

    {
        'bills': [...],
        'next': '...',
        'update': '/api.php/update'
    }

The only change is the addition of an *update* field, which works as such:

- When it is provided (i.e. not-null), the client should issue an AJAX **POST** on that URL to trigger an update on the server.
- When it is null, the client should ignore it.

I haven't pushed anything related to updates yet, but I've got the basics of timing and script execution down on the PHP side. I'll probably get them up after I eat - the integration work shouldn't be too difficult.

**EDIT** Now that I think about it, what are your thoughts on the UI of bill submission? The server should be able to provide some basics on the pending bills that it acquires via scraping (CBO and PDF URLs at the minimum, but usually its code, title, committee, Congress and publish date as well).

**Comment by markjm** [comment]
Sounds good on the updated JSONs. To be clear, I don't send any parameters back in the POST, correct?

How about any info you can't scrape, send some sentinel value in its place. On the client side, i can have something show like "We don't know what this value is, would you like to update this value?" Doing this would require a more full-fledged implementation of some POST commands.

I will spend Monday and Tuesday getting all the client side redone and ready to go. After that, I am happy to help add features to the backend. As a warning, though, I have a lot of trouble working with other's code - but I'm always looking to help.

**Comment by adamnew123456** [comment]
POST: Yeah, the backend provides the update URL in its entirety, you shouldn't need to do anything but POST to the URL in the response.

Also, those changes are running on the server if you want to look at it. I've set the update timeout to happen every 24 hours - if you need to test running the update script, let me know and I can tune the parameter down for a while.

Scraping: Currently any non-scrapable fields are represented as NULL in the DB, and I can represent them the same way in the JSON. It would probably be a good idea to make even provided fields editable - the summaries that it pulls are usually a little rough.

As for formatting of pending bills, they'll probably go to the frontend in the same format as normal bills do (excepting finance information), and the frontend can post them back to the backend using the same format.

**Comment by markjm** [comment]
Hey, all my ajax requests are coming back with a 426null response

Edit: Just kidding, I messed up.

**Comment by markjm** [comment]
Check out everything when you get a chance, and let me know if you see any problems.

Only thing I havent done yet is get it all set up to make edits.

**Comment by adamnew123456** [comment]
Looks good! Two small things:

1. It would be good if the UI could make the numbers a bit easier to read. Maybe something like adding commas to the numbers, or transforming "3900000000" into "3.9 billion".
2. The `format=json` parameter isn't used by the server. The server doesn't choke on it, but it might be worth removing depending how much code you devote to it.

**EDIT** Also, I have my schedule, and we're good to go for our 2:10 slot on Friday.

**Comment by adamnew123456** [comment]
Also, I just pushed some code for pulling down the list of pending bills - you can get a look at what they look like via `GET /api.php/bills/pending`. The only things of note are that finance entries are not present, sorting is disabled (results are always returned in oldest-first order to encourage cleaning up the backlog of unreviewed legislation - this also means that the `order` key is **not required**), and the 'update' key isn't present.

I haven't gotten submission working yet, I'll probably get that in tomorrow.

**Comment by adamnew123456** [comment]
I've done a bit of testing, and I think I've got submission working on my testing server - the bills I've submitted show up on the main feed, and are removed from the contribute page (specifically, the APPROVAL Act summary and the "Bill to Name Federal Courthouses"). It should be on the main server now.

The API is pretty much as we discussed it - just **POST** the bill in JSON form to the server at `/api.php/bills` and it will put it on the main feed, as well as remove its pending entry.
